Multiple issues at the same time

JanDeDobbeleer opened the pull request:
### Prerequisites

- [x] I have read and understood the [contributing guide][CONTRIBUTING.md].
- [x] The commit message follows the [conventional commits][cc] guidelines.
- [x] Tests for the changes have been added (for bug fixes / features).
- [x] Docs have been added/updated (for bug fixes / features).

### Description

- **fix(bash): correctly align multiline prompt**
- **refactor: simpler prompt caching logic**

<!---

Tips:

If you're not comfortable with working with Git,
we're working a guide (https://ohmyposh.dev/docs/contributing/git) to help you out.
Oh My Posh advises GitKraken (https://www.gitkraken.com/invite/nQmDPR9D) as your preferred cross platform Git GUI power tool.

-->

[CONTRIBUTING.md]: https://github.com/JanDeDobbeleer/oh-my-posh/blob/main/CONTRIBUTING.md
[cc]: https://www.conventionalcommits.org/en/v1.0.0/#summary

JanDeDobbeleer committed:
fix(bash): correctly align multiline prompt

relates to #5308

JanDeDobbeleer committed:
refactor: simpler prompt caching logic

JanDeDobbeleer commented:
@lewis-yeung if you find a moment, care to review the refactor commit?

lewis-yeung commented on src/prompt/engine.go:
```diff
@@ -510,32 +512,38 @@ func (e *Engine) adjustTrailingDiamondColorOverrides() {
 }
 
-func (e *Engine) updatePromptCache(value *promptCache) {
+func (e *Engine) updatePromptCache(value *engineCache) {
```
Should it also be renamed to `updateEngineCache` correspondingly?

lewis-yeung commented on src/prompt/primary.go:
```diff
@@ -139,4 +119,15 @@ func (e *Engine) writePrimaryRPrompt() {
 	e.write(e.rprompt)
 	e.write(terminal.RestoreCursorPosition())
+
+	if e.cached {
```
This will never be `true`, since an assignment is missing after the engine is restored from cache.

JanDeDobbeleer replied:
Good catch, it was there but I missed it after another refactor.

lewis-yeung commented on src/prompt/engine.go:
```diff
@@ -510,32 +512,38 @@ func (e *Engine) adjustTrailingDiamondColorOverrides() {
-func (e *Engine) checkPromptCache() bool {
-	data, ok := e.Env.Cache().Get(cache.PROMPTCACHE)
+func (e *Engine) restoreEngineFromCache() bool {
+	if !e.Env.Flags().Cached {
```
This flag is for the primary prompt only. Once a tooltip is rendered, we always cache it for reuse, regardless of this flag. Since we don't pass a `true` to this flag when requesting a tooltip from OMP CLI, if you check it here, a tooltip will never be cached.

JanDeDobbeleer replied:
But should we do that for PowerShell then?

lewis-yeung replied:
For PowerShell only? Yes, at least currently this approach of prompt caching is used in PowerShell only.

JanDeDobbeleer replied:
I would let the CLI determine this rather than a check in the code. That way it's reusable whenever.

lewis-yeung commented on src/prompt/primary.go:
```diff
@@ -139,4 +119,15 @@ func (e *Engine) writePrimaryRPrompt() {
+		CurrentLineLength: e.currentLineLength,
+		RPrompt:           e.rprompt,
+		RPromptLength:     e.rpromptLength,
+	})
```
How about leaving this part in the `Primary` method, and before its return?

```go
if !useCache {
  e.updatePromptCache(&engineCache{
    Prompt:            e.prompt.String(),
    CurrentLineLength: e.currentLineLength,
    RPrompt:           e.rprompt,
    RPromptLength:     e.rpromptLength,
  })
}

return e.string()
```

JanDeDobbeleer replied:
Agreed, it's more explicit that way.

lewis-yeung commented on src/cache/cache.go:
```diff
@@ -33,7 +33,7 @@ var (
 	TOGGLECACHE      = fmt.Sprintf("toggle_cache_%s", pid())
 	PROMPTCOUNTCACHE = fmt.Sprintf("prompt_count_cache_%s", pid())
-	PROMPTCACHE      = fmt.Sprintf("prompt_cache_%s", pid())
+	ENGINECACHE      = fmt.Sprintf("prompt_cache_%s", pid())
```
```suggestion
	ENGINECACHE      = fmt.Sprintf("engine_cache_%s", pid())
```

JanDeDobbeleer committed:
fix(bash): remove rprompt support

BREAKING CHANGE: rprompt is no longer supported for bash due to too many
side effects that have no solution in sight.

lewis-yeung commented on src/shell/scripts/omp.bash:
```diff
@@ -78,14 +78,8 @@ function _omp_hook() {
-# rprompt
-function _omp_rprompt() {
-    ::OMP:: print right --config="$POSH_THEME" --shell=bash --shell-version="$BASH_VERSION" --status="$_omp_status_cache" --pipestatus="${_omp_pipestatus_cache[*]}" --execution-time="$_omp_elapsed" --stack-count="$_omp_stack_count" --no-status="$_omp_no_exit_code" --terminal-width="${COLUMNS-0}" | tr -d '\0'
-    return $_omp_status_cache
```
This line should be added to the bottom of the `_omp_hook` function (where it was originally located).

JanDeDobbeleer replied:
@lewis-yeung I was unaware it was moved, again, good catch.

lewis-yeung replied:
I remember this because it was moved by me. 😆

lewis-yeung commented:
Couldn't agree more. The maintenance can be much less painful after completely removing the right prompt support for Bash. 😆

JanDeDobbeleer committed:
fix(install): validate permissions before upgrade

resolves #5318

JanDeDobbeleer committed:
feat(upgrade): do not upgrade to major versions automatically

resolves #5273

JanDeDobbeleer committed:
feat(font): cache font list

resolves #5300
